### ISSUE TITLE
race condition in closure-bound local variable handling

### ISSUE THREAD

**Issue by davidnich** [qorelanguage/qore#1265]
could be related to #1184, most likely the same issue as #1084

```
+ OK: "FtpClient Test" assertion at ./FtpClient/FtpClient.qtest:397
Assertion failed: (!rc), function unlock, file ../include/qore/QoreThreadLock.h, line 109.
Process 72757 stopped
* thread #3: tid = 0x63915b, 0x00007fff9ab27f06 libsystem_kernel.dylib`__pthread_kill + 10, stop reason = signal SIGABRT
    frame #0: 0x00007fff9ab27f06 libsystem_kernel.dylib`__pthread_kill + 10
libsystem_kernel.dylib`__pthread_kill:
->  0x7fff9ab27f06 <+10>: jae    0x7fff9ab27f10            ; <+20>
    0x7fff9ab27f08 <+12>: movq   %rax, %rdi
    0x7fff9ab27f0b <+15>: jmp    0x7fff9ab227cd            ; cerror_nocancel
    0x7fff9ab27f10 <+20>: retq   
(lldb) bt
* thread #3: tid = 0x63915b, 0x00007fff9ab27f06 libsystem_kernel.dylib`__pthread_kill + 10, stop reason = signal SIGABRT
  * frame #0: 0x00007fff9ab27f06 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff8b71c4ec libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00007fff932236df libsystem_c.dylib`abort + 129
    frame #3: 0x00007fff931eadd8 libsystem_c.dylib`__assert_rtn + 321
    frame #4: 0x0000000100162b22 libqore.5.dylib`QoreThreadLock::unlock(this=0x0000000101ceb048) + 82 at QoreThreadLock.h:109
    frame #5: 0x000000010039eed8 libqore.5.dylib`AutoLocker::~AutoLocker(this=0x00007000002095a0) + 24 at QoreThreadLock.h:159
    frame #6: 0x00000001000e0105 libqore.5.dylib`AutoLocker::~AutoLocker(this=0x00007000002095a0) + 21 at QoreThreadLock.h:158
    frame #7: 0x000000010028cafa libqore.5.dylib`qore_var_rwlock_priv::rdlock(this=0x0000000101ceb040) + 266 at qore_var_rwlock_priv.h:121
    frame #8: 0x000000010028c9e8 libqore.5.dylib`QoreVarRWLock::rdlock(this=0x0000000101ceaf68) + 24 at QoreRWLock.cpp:66
    frame #9: 0x000000010039deb3 libqore.5.dylib`QoreAutoVarRWReadLocker::QoreAutoVarRWReadLocker(this=0x0000700000209680, n_l=0x0000000101ceaf68) + 35 at qore_var_rwlock_priv.h:166
    frame #10: 0x00000001000d3d7d libqore.5.dylib`QoreAutoVarRWReadLocker::QoreAutoVarRWReadLocker(this=0x0000700000209680, n_l=0x0000000101ceaf68) + 29 at qore_var_rwlock_priv.h:165
    frame #11: 0x00000001000a1a4b libqore.5.dylib`ClosureVarValue::deref(this=0x0000000101ceaf60, xsink=0x0000700000209c80) + 491 at Variable.cpp:1330
    frame #12: 0x000000010038dda8 libqore.5.dylib`VarRefImmediateNode::derefImpl(this=0x0000000101cec330, xsink=0x0000700000209c80) + 40 at VarRefNode.h:326
    frame #13: 0x000000010004a0f5 libqore.5.dylib`AbstractQoreNode::deref(this=0x0000000101cec330, xsink=0x0000700000209c80) + 533 at AbstractQoreNode.cpp:153
    frame #14: 0x000000010006dba2 libqore.5.dylib`lvalue_ref::del(this=0x0000000101cec3f0, xsink=0x0000700000209c80) + 50 at qore_thread_intern.h:578
    frame #15: 0x000000010006db61 libqore.5.dylib`ReferenceNode::derefImpl(this=0x0000000101cec3d0, xsink=0x0000700000209c80) + 33 at ReferenceNode.cpp:214
    frame #16: 0x000000010004a0f5 libqore.5.dylib`AbstractQoreNode::deref(this=0x0000000101cec3d0, xsink=0x0000700000209c80) + 533 at AbstractQoreNode.cpp:153
    frame #17: 0x000000010004dc58 libqore.5.dylib`discard(n=0x0000000101cec3d0, xsink=0x0000700000209c80) + 40 at QoreLib.h:324
    frame #18: 0x00000001000e02ec libqore.5.dylib`VarValueBase::del(this=0x0000000101cec680, xsink=0x0000700000209c80) + 44 at LocalVar.h:114
    frame #19: 0x00000001000a19bf libqore.5.dylib`ClosureVarValue::deref(this=0x0000000101cec5d0, xsink=0x0000700000209c80) + 351 at Variable.cpp:1320
    frame #20: 0x000000010038dda8 libqore.5.dylib`VarRefImmediateNode::derefImpl(this=0x0000000100f7d0c0, xsink=0x0000700000209c80) + 40 at VarRefNode.h:326
    frame #21: 0x000000010004a0f5 libqore.5.dylib`AbstractQoreNode::deref(this=0x0000000100f7d0c0, xsink=0x0000700000209c80) + 533 at AbstractQoreNode.cpp:153
    frame #22: 0x000000010038e094 libqore.5.dylib`IntermediateParseReferenceNode::derefImpl(this=0x0000000100f7d160, xsink=0x0000700000209c80) + 52 at ParseReferenceNode.h:116
    frame #23: 0x000000010004a0f5 libqore.5.dylib`AbstractQoreNode::deref(this=0x0000000100f7d160, xsink=0x0000700000209c80) + 533 at AbstractQoreNode.cpp:153
    frame #24: 0x000000010025403c libqore.5.dylib`QoreListNode::derefImpl(this=0x0000000100f7a3e0, xsink=0x0000700000209c80) + 108 at QoreListNode.cpp:673
    frame #25: 0x000000010004a0f5 libqore.5.dylib`AbstractQoreNode::deref(this=0x0000000100f7a3e0, xsink=0x0000700000209c80) + 533 at AbstractQoreNode.cpp:153
    frame #26: 0x000000010037300c libqore.5.dylib`AbstractFunctionCallNode::~AbstractFunctionCallNode(this=0x0000000100f7d190) + 76 at FunctionCallNode.h:93
    frame #27: 0x0000000100382055 libqore.5.dylib`AbstractMethodCallNode::~AbstractMethodCallNode(this=0x0000000100f7d190) + 21 at FunctionCallNode.h:218
    frame #28: 0x000000010045a467 libqore.5.dylib`SelfFunctionCallNode::~SelfFunctionCallNode(this=0x0000000100f7d190) + 55 at FunctionCallNode.h:370
    frame #29: 0x00000001003590d5 libqore.5.dylib`SelfFunctionCallNode::~SelfFunctionCallNode(this=0x0000000100f7d190) + 21 at FunctionCallNode.h:369
    frame #30: 0x00000001003590f8 libqore.5.dylib`SelfFunctionCallNode::~SelfFunctionCallNode(this=0x0000000100f7d190) + 24 at FunctionCallNode.h:369
    frame #31: 0x000000010004a11d libqore.5.dylib`AbstractQoreNode::deref(this=0x0000000100f7d190, xsink=0x0000700000209eb0) + 573 at AbstractQoreNode.cpp:154
    frame #32: 0x0000000100285d8d libqore.5.dylib`BGThreadParams::exec(this=0x0000000100f7d270, xsink=0x0000700000209eb0) + 61 at thread.cpp:697
    frame #33: 0x0000000100285a94 libqore.5.dylib`::op_background_thread(x=0x0000000100f7d270) + 388 at thread.cpp:1933
    frame #34: 0x00007fff8b71999d libsystem_pthread.dylib`_pthread_body + 131
    frame #35: 0x00007fff8b71991a libsystem_pthread.dylib`_pthread_start + 168
    frame #36: 0x00007fff8b717351 libsystem_pthread.dylib`thread_start + 13
(lldb) fr s 11
frame #11: 0x00000001000a1a4b libqore.5.dylib`ClosureVarValue::deref(this=0x0000000101ceaf60, xsink=0x0000700000209c80) + 491 at Variable.cpp:1330
   1327 
   1328    while (true) {
   1329       {
-> 1330          QoreAutoVarRWReadLocker al(rml);
   1331 
   1332          if (!rset) {
   1333             if (ref_copy == rcount)
(lldb) p *this
error: xml-api-0.19.qmod debug map object file '/Users/david/src/qore/git/module-xml/src/.libs/single-compilation-unit.o' has changed (actual time is 0x57b4a73e, debug map time is 0x578dbbe9) since this executable was linked, file will be ignored
(ClosureVarValue) $0 = {
  VarValueBase = {
    val = {
      v = {
        b = false
        i = -256
        f = NaN
        n = 0xffffffffffffff00
      }
      type = '\0'
      fixed_type = true
      assigned = false
      static_assignment = false
    }
    id = 0x00000001031e93c1 "dataquit"
    skip = false
    finalized = false
  }
  RObject = {
    rml = {
      QoreVarRWLock = {
        priv = 0x0000000101ceb040
      }
    }
    tRefs = (references = 1)
    rlck = {
      ptm_lock = (__sig = 0, __opaque = char [56] @ 0x00007fa5126ede40)
    }
    rdone = {
      c = (__sig = 0, __opaque = char [40] @ 0x00007fa5126ede80)
    }
    rscan = 0
    rcount = 0
    rwaiting = 0
    rcycle = 0
    rset = 0x0000000000000000
    references = 0x0000000101ceb02c
    needs_is_valid = false
  }
  references = 0
  typeInfo = 0x00000001005888a0
}
```

**Comment by davidnich** [comment]
possibly fixed by the reference-counting fixes in #1268, #1270 and #1271 - TBE

**Comment by davidnich** [comment]
root cause #1084
